File: Merge Two Sorted Lists.cpp
//fail
//혼동했던 사실.
//무지성으로 temp = list1이니까 temp->next하면 list1->next도 바뀐다고 생각하면 안됰!! 그림 그려가면서 생각하자~!

/*
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* temp = list1;
        temp->next = list2->next;
        cout << list1->next->val << ' ' << temp->next->val << ' ';
        입력
        l1: [1,2,4]
        l2: [1,3,4]
        
        //결과 3 3
        return temp;
    }
};
*/
//재귀로짠거
class Solution {
public:
	ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
		if(l1 == NULL){
			return l2;
		}
		if(l2 == NULL){
			return l1;
		}        
		if(l1->val < l2->val){
			l1->next = mergeTwoLists(l1->next, l2);
			return l1;
		}
		else{
			l2->next = mergeTwoLists(l1, l2->next);
			return l2;            
		}
	}
};

//반복문
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if(l1 == NULL)
            return l2;
        if(l2 == NULL)
            return l1;
        
        ListNode* newH = NULL;
        ListNode* newT = NULL;
        
        if(l1->val < l2->val){
            newH = l1;
            newT = l1;
            l1 = l1->next;
        }
        else{
            newH = l2;
            newT = l2;
            l2 = l2->next;
        }
        
        while(l1 != NULL and l2 != NULL){
            if(l1->val < l2->val){
                newT->next = l1;
                newT = newT->next;
                l1 = l1->next;
                
            }else{
                newT->next = l2;
                newT = newT->next;
                l2 = l2->next;
            }
        }
        newT->next = (l1 == NULL ? l2 : l1);
        return newH;
    }
};

File: kotlin/Merge Two Sorted Lists.kt
//기본 재귀 해법
class Solution {
    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
        if(list1 == null )
            return list2
        if(list2 == null)
            return list1
        if(list1.`val` < list2.`val`){
            list1.next = mergeTwoLists(list1.next, list2)
            return list1
        }
        list2.next = mergeTwoLists(list2.next, list1)
        return list2
    }
}

//이건 기본 풀인데 배울게 많아서 가져옴.
//!! operator: NPE(Null-Pointer-Exception) 발생시키는 operator. 참고: https://kotlinlang.org/docs/null-safety.html#the-operator
class Solution {
    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
        if(list1 == null && list2 == null){return null}
        if(list1 == null && list2 != null){return list2}
        if(list1 != null && list2 == null){return list1}
        
        var next1 = list1
        var next2 = list2
        var firstVal: Int
        
        if(next1!!.`val` >= next2!!.`val`){
            firstVal = next2?.`val`
            next2 = next2.next
        }else{
            firstVal = next1?.`val`
            next1 = next1.next
        }
        
        var next3 = ListNode(firstVal)
        var list3Start = next3

        while (next1!= null || next2!= null){
            if(next1== null){
                next3.next = next2
                break
            }
            if(next2== null){
                next3.next = next1
                break
            }
            if(next1!!.`val` >= next2!!.`val`){
                next3.next = ListNode(next2?.`val`)
                next3 = next3.next
                next2 = next2.next
            }else{
                next3.next = ListNode(next1?.`val`)
                next3 = next3.next
                next1 = next1.next 
            }
        }        
        
        return list3Start
    }
}
